third_party/e2e/engine.cc

```diff
@@ -0,0 +1,90 @@
+#include "engine.h"
+
+#include <cstdint>
+
+#include "lib/translator/translation.h"
+#include "nvme_driver.h"
+
+namespace {
+
+uint64_t AllocPagesCallback(uint16_t count) { return AllocPages(count); }
+
+void DeallocPagesCallback(uint64_t addr, uint16_t count) {
+  DeallocPages(addr, count);
+}
+
+}  // namespace
+
+void SetEngineCallbacks(void) {
+  translator::SetDebugCallback(Print);
+  translator::SetAllocPageCallbacks(AllocPagesCallback, DeallocPagesCallback);
+}
+
+ScsiToNvmeResponse ScsiToNvme(unsigned char* cmd_buf, unsigned short cmd_len,
+                              unsigned long long lun, unsigned char* sense_buf,
+                              unsigned short sense_len, unsigned char* data_buf,
+                              unsigned short data_len, bool is_data_in) {
+  // Create translation object
+  translator::Translation translation;
+
+  // Package parameters and run translation begin
+  translator::Span<uint8_t> scsi_cmd(cmd_buf, cmd_len);
+  translator::Span<uint8_t> buffer(data_buf, data_len);
+  translator::BeginResponse begin_resp =
+      translation.Begin(scsi_cmd, buffer, lun);
+
+  if (begin_resp.status == translator::ApiStatus::kFailure) {
+    Print("Incorrect usage of Translation Library API");
+    ScsiToNvmeResponse resp = {.return_code = 0x40, .alloc_len = 0};
+    return resp;
+  }
+
+  if (begin_resp.alloc_len > data_len) {
+    Print(
+        "Specified allocation length exceeds buffer size. Possible malicious "
+        "request?");
+    ScsiToNvmeResponse resp = {.return_code = 0x40, .alloc_len = 0};
+    return resp;
+  }
+
+  // Grab NVMe cmds and call NVMe interface
+  translator::Span<const translator::NvmeCmdWrapper> nvme_wrappers =
+      translation.GetNvmeWrappers();
+  nvme::GenericQueueEntryCpl cpl_buf[nvme_wrappers.size()] = {};
+  for (uint32_t i = 0; i < nvme_wrappers.size(); ++i) {
+    NvmeCommand tmp_cmd;
+    NvmeCompletion tmp_cpl = {};
+    memcpy(&tmp_cmd, &nvme_wrappers[i].cmd, sizeof(tmp_cmd));
+    static_assert(sizeof(tmp_cmd) == sizeof(nvme_wrappers[i].cmd));
+    void* buffer = reinterpret_cast<void*>(nvme_wrappers[i].cmd.dptr.prp.prp1);
+    unsigned bufflen = 4096;
```

Comment:
Is there a reason we don't want to specify the type here? To use int or long or something more specific